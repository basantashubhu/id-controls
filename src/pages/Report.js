import firebase from '../firebase'
import {useEffect, useState} from "react";

const Report = ({location}) => {
    const [diseases, setDiseases] = useState([]);
    const [addresses, setAddresses] = useState([]);
    const [formData, setFormData] = useState({name : '', address : '', disease : ''})
    const [message, setMessage] = useState({type : 'danger', message : ''})
    const [loading, setLoading] = useState(false)
    useEffect(() => {
        return firebase.firestore().collection('diseases').onSnapshot(snap => {
            setDiseases([...new Set(snap.docs.map(d => d.data()))]);
        });
    }, [diseases.length])
    useEffect(() => {
        return firebase.firestore().collection('cases').onSnapshot(snap => {
            setAddresses([...new Set(snap.docs.map(c => c.data().address))]);
        });
    },[addresses.length])
    const handleInputChange = e => {
        setFormData({
            ...formData,
            [e.target.name] : e.target.value
        });
    };
    const handleFormSubmit = async e => {
        e.preventDefault();
        try {
            setLoading(true)
            setMessage({})
            await firebase.firestore().collection('cases').add(formData)
            setLoading(false)
            setFormData({name : '', address : '', disease : ''})
            setMessage({type : 'success', message : 'Your case is submitted successfully.'})
        } catch (e) {
            setLoading(false)
            setMessage({type : 'danger', message : e.message})
        }
    };
    const search = new URLSearchParams(location.search),
        id = search.get('disease') || '';
    useEffect(() => {
        setFormData({...formData, disease : id})
    }, [id, formData])
    return <main>
        <section className="container">
            <div className="row py-lg-5">
                <div className="col-lg-6 col-md-8 mx-auto">
                    <h1 className="fw-light">Report a case</h1>
                    <form className="mt-5" onSubmit={handleFormSubmit}>
                        <div className={`alert alert-${message.type}`} hidden={!message.message}>{message.message}</div>
                        <div className="form-group">
                            <label>Name</label>
                            <input type="text" name="name" className="form-control"
                                   required={true} onChange={handleInputChange} value={formData.name}/>
                        </div>
                        <div className="form-group">
                            <label>Address</label>
                            <input type="text" name="address" className="form-control" list="case-addresses"
                                   required={true} onChange={handleInputChange} value={formData.address}/>
                            <datalist id="case-addresses">
                                {addresses.map((addr,index) => <option value={addr} key={index}/>)}
                            </datalist>
                        </div>
                        <div className="form-group">
                            <label>Disease</label>
                            <select name="disease" className="form-control"
                                    required={true} onChange={handleInputChange} value={formData.disease}>
                                <option value="">Choose</option>
                                {diseases.map((d, index) => {
                                    return <option value={d.id} key={index}>{d.disease}</option>
                                })}
                            </select>
                        </div>
                        <div className="form-group mt-3">
                            <button type="submit" className="btn btn-primary" disabled={loading}>{loading ? 'Submitting...' : 'Submit'}</button>
                        </div>
                    </form>
                </div>
            </div>
        </section>
    </main>
}

export default Report