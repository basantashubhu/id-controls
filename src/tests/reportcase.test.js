import React from 'react'
import {shallow, configure} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Report from '../pages/Report'
configure({adapter :  new Adapter()})
describe('Test case for contact form', () => {
    let wrapper, location = {search : ''};
    it('name check', () => {
        wrapper = shallow(<Report location={location}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('[name="name"]')).toHaveLength(1);
    })
    it('address check', () => {
        wrapper = shallow(<Report location={location}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('[name="address"]')).toHaveLength(1);
    })
    it('disease check', () => {
        wrapper = shallow(<Report location={location}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('[name="disease"]')).toHaveLength(1);
    })
    it('submit button check', () => {
        wrapper = shallow(<Report location={location}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('[type="submit"]')).toHaveLength(1);
    })
})
