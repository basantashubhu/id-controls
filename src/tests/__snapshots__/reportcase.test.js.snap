// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for contact form address check 1`] = `ShallowWrapper {}`;

exports[`Test case for contact form disease check 1`] = `ShallowWrapper {}`;

exports[`Test case for contact form name check 1`] = `ShallowWrapper {}`;

exports[`Test case for contact form submit button check 1`] = `ShallowWrapper {}`;
